design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,181 @@
+
+   **Pros:**
+   - The upgrade-specific code can be removed after two or three releases.
+   - Simple logic after the upgrade is completed.
+   - Limited disturbance to git-ops workflows, affecting only pre-existing VMs with the current method.
+
+   **Cons:**
+   - Requires additional code (potentially in `virt-operator`) to handle the upgrade-specific logic.
+
+
+
+### Phase 2: Transition to Universally Unique UUID Generation
+
+In the second phase, the approach for generating the firmware UUID will be modified to ensure global uniqueness.
+This transition could include changing the UUID generation to consider both the VM name and namespace,
+or adopting a pseudo-random hash to guarantee uniqueness across clusters. The solutions for this phase are outlined below:
+
+1. **Use Namespace + Name for UUID Generation**  
+   **Description:** Generate the UUID by combining the VM name and namespace, ensuring uniqueness within the cluster.
+
+   **Pros:**
+   - Simple and straightforward to implement.
+   - Avoids reliance on an external generator for randomness.
+
+   **Cons:**
+   - Introduces a breaking change, as UUIDs for existing VMs may change unless preserved.
+
+
+2. **Adopt Pseudo-Random Hashing for UUID Generation**  
+   **Description:** Use a pseudo-random hash to generate a unique UUID that does not rely on specific names or namespaces.
+
+   **Pros:**
+   - Ensures global uniqueness, even across clusters.
+   - Avoids dependence on VM names and namespaces, providing more flexibility.
+
+   **Cons:**
+   - Introducing randomness could make the UUID less predictable, which may impact certain workloads.
+   - Requires careful validation to ensure the hash remains consistent for each VM across restarts.
+
+
+3. **Use `vm.metadata.uid` as the Firmware UUID**  
+   **Description:** Use the `vm.metadata.uid`, a unique identifier assigned to each VM by Kubernetes, as the firmware UUID.
+
+   **Pros:**
+   - Ensures uniqueness within and across clusters, as `metadata.uid` is universally unique for each VM.
+   - Avoids adding new fields or complex generation logic.
+   - Simplifies UUID management by leveraging an existing unique identifier.
+
+   **Cons:**
+
+
+By addressing each phase separately, we can balance immediate workload stability with a gradual transition to globally unique UUIDs. This approach allows for flexibility in implementation and can minimize disruption for end-users.
+
+
+## API Examples
+**Example uuid persistence within VM Status**
+
+```yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: mytestvm
+status:
+   conditions:
+      - lastProbeTime: "2024-11-06T01:12:29Z"
+        lastTransitionTime: "2024-11-06T01:12:29Z"
+        message: Guest VM is not reported as running
+        reason: GuestNotRunning
+        status: "False"
+        type: Ready
+   created: true
+   runStrategy: Once
+   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
+```
+
+or persist via status condition
```

Comment:
I'd argue against a condition. Conditions are generally needed to manage the lifecycle or to be consumed by other controllers.

Reply:
conditions are not for persistence.